packages/lesswrong/components/ultraFeed/FeedContentBody.tsx

```diff
@@ -0,0 +1,277 @@
+  
+  const isMaxLevel = expansionLevel >= breakpoints.length - 1;
+  const currentWordLimit = breakpoints[expansionLevel];
+  const isFirstLevel = expansionLevel === 0;
+  
+  const documentType = post ? 'post' : comment ? 'comment' : 'tag';
+  const documentId = post?._id || comment?._id || tag?._id;
+  
+  const getDocumentUrl = () => {
+    if (post) return postGetPageUrl(post);
+    if (comment) return commentGetPageUrlFromIds({
+      postId: comment.postId,
+      commentId: comment._id
+    });
+    if (tag) return `/tag/${tag._id}`;
+    return '/';
+  };
+  
+  const handleExpand = useCallback(() => {
+    if (isMaxLevel && linkToDocumentOnFinalExpand) {
+        return;
+    }
+
+    const newLevel = Math.min(expansionLevel + 1, breakpoints.length - 1);
+    const newMaxReached = newLevel >= breakpoints.length - 1;
+    setExpansionLevel(newLevel);
+    onExpand?.(newLevel, newMaxReached, wordCount);
+
+  }, [
+    expansionLevel,
+    breakpoints.length,
+    onExpand,
+    isMaxLevel,
+    linkToDocumentOnFinalExpand,
+    wordCount,
+  ]);
+  
+  // Handle clicks on the content area
+  const handleContentClick = useCallback((e: React.MouseEvent) => {
+    const target = e.target as HTMLElement;
+
+    // If the user clicking on a link, allow default navigation and don't expand.
+    if (target.tagName === 'A' || target.parentElement?.tagName === 'A') {
+      return;
+    }
+
+    // Otherwise, it's a click on general content. Trigger expansion.
+    handleExpand();
+
+  }, [handleExpand]); // Remove isFirstLevel from dependency array
+  
+  const usingLineClamp = clampOverride !== undefined && clampOverride > 0;
+  
+  // Process the HTML content only if not using line clamp
+  let truncatedHtml = html;
+  let wasTruncated = false;
+  let wordsLeft = 0;
+  
+  if (!usingLineClamp) {
+    const createReadMoreSuffix = () => {
+      if (isMaxLevel && linkToDocumentOnFinalExpand) {
+        return '...';
+      }
+      return `...<span class="read-more-button" data-expansion-level="${expansionLevel}">(read more)</span>`;
+    };
+    
+    const result = truncateWithGrace(
+      html,
+      currentWordLimit,
+      20,
+      wordCount,
+      createReadMoreSuffix()
+    );
+    
+    truncatedHtml = result.truncatedHtml;
+    wasTruncated = result.wasTruncated;
+    wordsLeft = result.wordsLeft;
+    
+    // Direct check to ensure we correctly identify if content is truncated
+    // This handles cases where truncateWithGrace might not accurately report truncation
+    if (wordCount > currentWordLimit && !wasTruncated) {
+      wasTruncated = true;
+      wordsLeft = wordCount - currentWordLimit;
+    }
+  } else {
+    // For line clamp mode, assume there's more content if word count exceeds a threshold
+    // (This is an estimate since we can't know exactly without rendering)
+    wasTruncated = wordCount > currentWordLimit;
+    wordsLeft = wordCount - currentWordLimit;
+  }
```

Comment:
This should be a function returning the three relevant values, rather than an inline if-block assigning to mutable variables